### PR TITLE
V2 of gentlemanjerry

### PR DESCRIPTION
**Pull request by almathew**

### CODE REVIEW

**Comment by almathew** [comment]
@UserNotFound do you think this is too risky? (I can't add a comment to the line because I didn't edit it...)
```# We're using Redis to route logs under pub / sub.  No use saving anything, so
# we'll disable RDB and AOF persistence.
save ""
appendonly no
```

File: Dockerfile.erb [hunk 1]
```diff
@@ -34,12 +34,13 @@ RUN ./install-stunnel.sh
...
 # Update http output gem
 # Install our syslog output implementation
-RUN apk add --no-cache git && \
+RUN apk add --no-cache git sed && \
     GEMFILE="/logstash-${LOGSTASH_VERSION}/Gemfile" && \
     grep -v 'logstash-output-http' "$GEMFILE" > "${GEMFILE}.tmp" && \
     mv "${GEMFILE}.tmp" "$GEMFILE" && \
     grep -v 'logstash-output-redis' "$GEMFILE" > "${GEMFILE}.tmp" && \
     mv "${GEMFILE}.tmp" "$GEMFILE" && \
+    sed -i 's#https:\/\/rubygems#http:\/\/rubygems#g' $GEMFILE && \
```
Comment (almathew) [inline Dockerfile.erb]: This is pretty unacceptable, and I'm still working on a fix. See the slack convo here for what has been tried: https://aptible.slack.com/archives/C02NMLZ8N/p1574263863001400

**Comment by almathew** [comment]
Ok, so covering the last issue I had pending: the core issue that jruby-openssl 0.9.7 (which is what Logstash comes bundled with) doesn't support whatever Rubygems.org is using for security. Our build failures line up with this: https://blog.rubygems.org/2018/02/24/tls-10-and-11-deprecation-notice.html

@UserNotFound handing this change off to you for the final 👍  since you and @joshraker will be supporting it

I'm getting around this by upgrading to 0.9.20. Unfortunately I can't use the Logstash-packaged version to do the upgrade, since it can't connect to rubygems.org :tada:
So instead the solution is to install jruby, upgrade the package, and uninstall jruby.

cc @fancyremarker 
(1) because I'm a little confused about how this change plays into this: https://github.com/aptible/gentlemanjerry/pull/38
and (2) on the off chance that you have a better idea (although please don't put a ton of time into thinking about it)

File: Dockerfile.erb [hunk 2]
```diff
@@ -32,6 +32,13 @@ RUN apk add --no-cache coreutils redis
 ADD bin/install-stunnel.sh install-stunnel.sh
 RUN ./install-stunnel.sh
 
+# Use another version of jruby to upgrade jruby-openssl,
+# since the Logstash-packaged version of jruby-openssl can't connect
+# rubygems.org
+RUN apk add jruby && \
```
Comment (fancyremarker) [inline Dockerfile.erb]: To make sure I understand:

Is the other version of JRuby is installed as part of the Logstash installation package above?

```
RUN curl -O "https://download.elastic.co/logstash/logstash/logstash-${LOGSTASH_VERSION}.tar.gz" && \ # ...
```

And, can we check that the `docker diff` of running this step on top of the previous step does not inadvertently delete any of the Logstash-installed JRuby files?

Assuming so, and assuming this then works, I think this is a clever and good fix. As long as the answer to my second question is yes, this should not affect https://github.com/aptible/gentlemanjerry/pull/38.

Comment (fancyremarker) [reply Dockerfile.erb]: In trying to build this locally via `TAG=large make build`, I hit:

```
Error: Your application used more memory than the safety cap of 500M.
Specify -J-Xmx####m to increase it (#### = cap size in MB).
Specify -w for full OutOfMemoryError stack trace
```

Is there something extra/different I should to in order to build locally?

Comment (almathew) [reply Dockerfile.erb]: Diff (and `gem list`) confirms that nothing was deleted. It's 90% additions, with a few changes.

This is probably irrelevant to your error, but I got rid of the tags (managing that has been moved to sweetness), although I do see there's some stuff I missed in the Makefile now.

At which point in the process are you hitting that?

**Comment by UserNotFound** [comment]
> @UserNotFound do you think this is too risky? (I can't add a comment to the line because I didn't edit it...)
> 
> ```
> # we'll disable RDB and AOF persistence.
> save ""
> appendonly no
> ```

In my mind, this is the same risk as losing logs from within the java heap of logstash itself.  It would be nice if we had an alert (on container recovery kicking in?) to let us know when this occurs, otherwise it'll only manifest as missing logs, afaik.

I think if we find that this does bite us, it's easy enough to change as long as we consider the lifecycle of this container filesystem and the space available on the logging instance.

File: test/load-message.bats [hunk 3]
```diff
@@ -183,7 +183,7 @@ teardown() {
 
   # We'll send messages that are about 40kB. We should not be able to store
   # more than 50 of those, so we'll send 100.
-  n_messages=100
+  n_messages=50
```
Comment (UserNotFound) [inline test/load-message.bats]: Either this was a typo, or the explanation in the comment should be updated.

File: Dockerfile.erb (hunk 2 above)
Comment (UserNotFound) [reply Dockerfile.erb]: >although I do see there's some stuff I missed in the Makefile now.
`.travis.yml`, as well

Comment (almathew) [reply Dockerfile.erb]: travis.yml updated. That's what the problem was, not the Makefile.

**Pull request merged**